crates/circuit/src/dag_circuit.rs

```diff
@@ -0,0 +1,6474 @@
+    }
+
+    fn _insert_1q_on_incoming_qubit(
+        &mut self,
+        py: Python,
+        node: &Bound<PyAny>,
+        old_index: usize,
+    ) -> PyResult<()> {
+        if let NodeType::Operation(inst) = self.pack_into(py, node)? {
+            self.increment_op(inst.op.name().to_string());
+            let new_index = self.dag.add_node(NodeType::Operation(inst));
+            let old_index: NodeIndex = NodeIndex::new(old_index);
+            let (parent_index, edge_index, weight) = self
+                .dag
+                .edges_directed(old_index, Incoming)
+                .map(|edge| (edge.source(), edge.id(), edge.weight().clone()))
+                .next()
+                .unwrap();
+            self.dag.add_edge(parent_index, new_index, weight.clone());
+            self.dag.add_edge(new_index, old_index, weight);
+            self.dag.remove_edge(edge_index);
+            Ok(())
+        } else {
+            Err(PyTypeError::new_err("Invalid node type input"))
+        }
+    }
+
+    fn _edges(&self, py: Python) -> Vec<PyObject> {
+        self.dag
+            .edge_indices()
+            .map(|index| {
+                let wire = self.dag.edge_weight(index).unwrap();
+                match wire {
+                    Wire::Qubit(qubit) => self.qubits.get(*qubit).to_object(py),
+                    Wire::Clbit(clbit) => self.clbits.get(*clbit).to_object(py),
+                    Wire::Var(var) => var.clone_ref(py),
+                }
+            })
+            .collect()
+    }
+}
+
+impl DAGCircuit {
+    /// Return an iterator of gate runs with non-conditional op nodes of given names
+    pub fn collect_runs(
+        &self,
+        namelist: HashSet<String>,
+    ) -> Option<impl Iterator<Item = Vec<NodeIndex>> + '_> {
+        let filter_fn = move |node_index: NodeIndex| -> Result<bool, Infallible> {
+            let node = &self.dag[node_index];
+            match node {
+                NodeType::Operation(inst) => Ok(namelist.contains(inst.op.name())
+                    && match &inst.extra_attrs {
+                        None => true,
+                        Some(attrs) => attrs.condition.is_none(),
+                    }),
+                _ => Ok(false),
+            }
+        };
+        rustworkx_core::dag_algo::collect_runs(&self.dag, filter_fn)
+            .map(|node_iter| node_iter.map(|x| x.unwrap()))
+    }
+
+    /// Return a set of non-conditional runs of 1q "op" nodes.
+    pub fn collect_1q_runs(&self) -> Option<impl Iterator<Item = Vec<NodeIndex>> + '_> {
+        let filter_fn = move |node_index: NodeIndex| -> Result<bool, Infallible> {
+            let node = &self.dag[node_index];
+            match node {
+                NodeType::Operation(inst) => Ok(inst.op.num_qubits() == 1
+                    && inst.op.num_clbits() == 0
+                    && inst.op.matrix(inst.params_view()).is_some()
+                    && inst.condition().is_none()),
+                _ => Ok(false),
```

Comment:
This always constructs a matrix, just to test whether it's _possible_ to construct a matrix.  We can at least short-circuit that for standard gates; all standard gates can create a matrix of themselves if the parameters are all floats.

Reply:
Yeah, I have a fix for this already on #12959 to make it: something like (although I should move the `is_parameterized()` outside the outer paren there.

```suggestion
                NodeType::Operation(inst) => Ok(inst.op.num_qubits() == 1
                    && inst.op.num_clbits() == 0
                    && ((!inst.is_parameterized() && inst.op.try_standard_gate(
).is_some())
                        || inst.op.matrix(inst.params_view()).is_some())
                    && inst.condition().is_none()),
                _ => Ok(false),
```

Reply:
Updated in: https://github.com/Qiskit/qiskit/pull/12550/commits/7e4307182aa2f4d626cbd60c2a35f9ecfc00bd07